docs/enhancements/1-crds-and-controllers.md

```diff
@@ -0,0 +1,430 @@
+| Status | Authors   | Created    | 
+|--------|-----------|------------|
+| Review | @jewertow | Jan 2 2025 |
+
+# CRDs and controllers
+
+## Overview
+
+Currently, we use helm to manage the controller deployment and its configuration. The configuration is passed to the controller as CLI arguments.
+This approach requires restarting the controller on any configuration change and causes pod failures in case of a misconfiguration.
+
+Additionally, current approach to resource management is not reliable. We apply resources when it's necessary using the Istio client's `Apply` function,
+so we are not able to reconcile created objects on user's modifications or controller uninstallation.
+
+### Goals
+
+Design CRDs for better user experience and ensure proper resource lifecycle and garbage collection.
+
+### Non-goals
+
+Design CRD for installing federation controller. Federation controller will be still installed with Helm,
+and we are not planning any dedicated operator or integration with Sail Operator for now.
+
+## Design
+
+We need the following CRDs:
+1. `MeshFederation` - cluster-scoped resource that includes general federation config, i.e. local settings, remote addresses and identities.
+2. `FederationServiceRules` - specifies rules for exporting and importing services; parent for export-related Istio resources, i.e. `Gateway`, `DestinationRule`, etc.; must be created by the mesh admin.
+3. `ImportedService` - represents an imported service; parent for import-related Istio resources, i.e. `ServiceEntry`, `WorkloadEntry`, etc.; managed only by the controller; should not be touch by users.
+
+All resources will contain [standard conditions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) in their status.
+
+#### MeshFederation
+
+`MeshFederation` includes settings related only to mesh federation, like remote ingress address, identities, etc.
+This resource must be cluster-scoped, because it will be used as an owner for `ImportedService` resources, which can be created in multiple namespaces.
+
+I would like to move away from thinking about "peers", as this is just an implementation detail.
+I also thought that it may be confusing why peer address is used as the address for remote services.
+
+Resource structure:
+
+```yaml
+apiGroup: federation.openshift-service-mesh.io/v1alpha1
+kind: MeshFederation
+metadata:
+  name: default
+spec:
+  local:
+    # ID is a unique identifier of the FDS peer, and it's used as a suffix in its Service and ServiceAccount names.
+    id: east
```

Comment:
I think we should use `.metadata.name` instead. Then we offload handling the case of rename (remove and recreate child resources) to k8s clean-up instead, because user will deliberately need to first delete the resource and then create under the new name. That would imply that the existing RemoteMeshes are recreated by the user, but I would think of it as analogous case to remove Namespace with Services in it. If you want to rename the ns you do delete/create instead.

Im my opinion we should keep it simple for now and avoid corner cases unless they're really necessary from the UX perspective.

Reply:
OK